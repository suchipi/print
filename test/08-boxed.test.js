const print = require("../print");

describe("Boxed primitives", () => {
	describe("Booleans", () => {
		it("prints internal values", () => {
			expect(print(new Boolean("Yes"))).toMatchSnapshot();
			expect(print(new Boolean(false))).toMatchSnapshot();
			expect(print(new Boolean())).toMatchSnapshot();
		});

		it("prints properties", () => {
			let value = new Boolean("Yes");
			expect(print(value)).toMatchSnapshot();
			value.foo = "Foo";
			expect(print(value)).toMatchSnapshot();

			value = new Boolean(false);
			expect(print(value)).toMatchSnapshot();
			value.bar = "Bar";
			expect(print(value)).toMatchSnapshot();
		});

		it("identifies subclasses", () => {
			class ExtendedBoolean extends Boolean {}
			const value = new ExtendedBoolean("Yes");
			expect(print(value)).toMatchSnapshot();
			value.foo = "Foo";
			expect(print(value)).toMatchSnapshot();
		});

		it("resolves internal values with `Boolean#valueOf()`", () => {
			const value = new Boolean("Yes");
			Object.defineProperties(value, {
				valueOf: {
					value() {
						return false;
					},
				},
				toString: {
					value() {
						return "Nah";
					},
				},
			});
			expect(print(value)).toMatchSnapshot();
			value.foo = "Foo";
			expect(print(value)).toMatchSnapshot();

			class WeirdBoolean extends Boolean {
				toString() {
					return "Nah";
				}
				valueOf() {
					return !super.valueOf();
				}
			}
			expect(print(new WeirdBoolean(true))).toMatchSnapshot();
			expect(print(new WeirdBoolean(false))).toMatchSnapshot();
		});
	});

	describe("Numbers", () => {
		it("prints internal values", () => {
			expect(print(new Number(64))).toMatchSnapshot();
			expect(print(new Number(-326.2))).toMatchSnapshot();
		});

		it("prints properties", () => {
			const value = new Number(48);
			value.foo = "Foo";
			expect(print(value)).toMatchSnapshot();
			value.bar = "Bar";
			expect(print(value)).toMatchSnapshot();
		});

		it("identifies subclasses", () => {
			class Double extends Number {
				constructor(n) {
					super(n * 2);
				}
			}
			class Quadruple extends Double {
				constructor(n) {
					super(n * 2);
				}
			}

			let value = new Double(150);
			expect(print(value)).toMatchSnapshot();
			value.foo = "Foo";
			expect(print(value)).toMatchSnapshot();

			value = new Quadruple(150);
			expect(print(value)).toMatchSnapshot();
			value.bar = "Bar";
			expect(print(value)).toMatchSnapshot();
		});

		it("identifies Math.* constants", () => {
			class MathConstant extends Number {}
			for (const constant of "E LN10 LN2 LOG10E LOG2E PI SQRT1_2 SQRT2".split(
				" "
			)) {
				let value = new Number(Math[constant]);
				expect(print(value)).toMatchSnapshot();
				value.foo = "Foo";
				expect(print(value)).toMatchSnapshot();

				value = new MathConstant(Math[constant]);
				expect(print(value)).toMatchSnapshot();
				value.bar = "Bar";
				expect(print(value)).toMatchSnapshot();
			}
		});

		it("identifies Number.* constants", () => {
			class MagicNumber extends Number {}
			const ants =
				"EPSILON MIN_VALUE MAX_VALUE MIN_SAFE_INTEGER MAX_SAFE_INTEGER".split(
					" "
				);
			for (const constant of ants) {
				let value = new Number(Number[constant]);
				expect(print(value)).toMatchSnapshot();
				value.foo = "Foo";
				expect(print(value)).toMatchSnapshot();

				value = new MagicNumber(Number[constant]);
				expect(print(value)).toMatchSnapshot();
				value.foo = "Foo";
				expect(print(value)).toMatchSnapshot();
			}
		});

		it("resolves internal values with `Number#valueOf()`", () => {
			let value = new Number(258);
			value.valueOf = function () {
				return 30;
			};
			expect(print(value)).toMatchSnapshot();
			value.toString = function () {
				return "982";
			};
			expect(print(value)).toMatchSnapshot();

			class LyingNumber extends Number {
				toString() {
					return "Nah";
				}
				valueOf() {
					return -752;
				}
			}
			value = new LyingNumber(32);
			expect(print(value)).toMatchSnapshot();
			value.foo = "Foo";
			expect(print(value)).toMatchSnapshot();
		});
	});

	describe("Strings", () => {
		it("prints internal values", () => {
			const value = new String("ABC");
			expect(print(value)).toMatchSnapshot();
		});

		it("prints properties", () => {
			const value = new String("XYZ");
			value.foo = "Foo";
			expect(print(value)).toMatchSnapshot();
		});

		it("identifies subclasses", () => {
			class ExtendedString extends String {}
			const value = new ExtendedString("XYZ");
			expect(print(value)).toMatchSnapshot();
			value.foo = "Foo";
			expect(print(value)).toMatchSnapshot();
		});

		it("escapes whitespace characters", () => {
			const escapeTests = [
				['"\\t"', new String("\t")],
				['"\\n"', new String("\n")],
				['"\\f"', new String("\f")],
			];
			class Whitespace extends String {}
			for (const [char, object] of escapeTests) {
				expect(print(object)).toMatchSnapshot(char);
				object.foo = "Foo";
				expect(print(object)).toMatchSnapshot(char);

				const ws = new Whitespace(object + "");
				expect(print(ws)).toMatchSnapshot(char);
				ws.bar = "Bar";
				expect(print(ws)).toMatchSnapshot(char);
			}
		});

		it("resolves internal values with `String#valueOf()`", () => {
			let value = new String("ABC");
			value.toString = function () {
				return "XYZ";
			};
			value.valueOf = function () {
				return "Nah";
			};
			expect(print(value)).toMatchSnapshot();

			class WeirdString extends String {
				toString() {
					return "XYZ";
				}
				valueOf() {
					return "XYZ";
				}
			}
			value = new WeirdString("ABC");
			expect(print(value)).toMatchSnapshot();
		});
	});
});
