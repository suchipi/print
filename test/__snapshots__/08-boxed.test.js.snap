// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Boxed primitives Booleans identifies subclasses 1`] = `
ExtendedBoolean {
	true
}
`;

exports[`Boxed primitives Booleans identifies subclasses 2`] = `
ExtendedBoolean {
	true
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Booleans prints internal values 1`] = `
Boolean {
	true
}
`;

exports[`Boxed primitives Booleans prints internal values 2`] = `
Boolean {
	false
}
`;

exports[`Boxed primitives Booleans prints internal values 3`] = `
Boolean {
	false
}
`;

exports[`Boxed primitives Booleans prints properties 1`] = `
Boolean {
	true
}
`;

exports[`Boxed primitives Booleans prints properties 2`] = `
Boolean {
	true
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Booleans prints properties 3`] = `
Boolean {
	false
}
`;

exports[`Boxed primitives Booleans prints properties 4`] = `
Boolean {
	false
	
	bar: "Bar"
}
`;

exports[`Boxed primitives Booleans resolves internal values with \`Boolean#valueOf()\` 1`] = `
Boolean {
	true
}
`;

exports[`Boxed primitives Booleans resolves internal values with \`Boolean#valueOf()\` 2`] = `
Boolean {
	true
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Booleans resolves internal values with \`Boolean#valueOf()\` 3`] = `
WeirdBoolean {
	true
}
`;

exports[`Boxed primitives Booleans resolves internal values with \`Boolean#valueOf()\` 4`] = `
WeirdBoolean {
	false
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 1`] = `
Number {
	Math.E
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 2`] = `
Number {
	Math.E
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 3`] = `
MathConstant {
	Math.E
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 4`] = `
MathConstant {
	Math.E
	
	bar: "Bar"
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 5`] = `
Number {
	Math.LN10
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 6`] = `
Number {
	Math.LN10
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 7`] = `
MathConstant {
	Math.LN10
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 8`] = `
MathConstant {
	Math.LN10
	
	bar: "Bar"
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 9`] = `
Number {
	Math.LN2
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 10`] = `
Number {
	Math.LN2
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 11`] = `
MathConstant {
	Math.LN2
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 12`] = `
MathConstant {
	Math.LN2
	
	bar: "Bar"
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 13`] = `
Number {
	Math.LOG10E
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 14`] = `
Number {
	Math.LOG10E
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 15`] = `
MathConstant {
	Math.LOG10E
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 16`] = `
MathConstant {
	Math.LOG10E
	
	bar: "Bar"
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 17`] = `
Number {
	Math.LOG2E
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 18`] = `
Number {
	Math.LOG2E
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 19`] = `
MathConstant {
	Math.LOG2E
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 20`] = `
MathConstant {
	Math.LOG2E
	
	bar: "Bar"
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 21`] = `
Number {
	Math.PI
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 22`] = `
Number {
	Math.PI
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 23`] = `
MathConstant {
	Math.PI
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 24`] = `
MathConstant {
	Math.PI
	
	bar: "Bar"
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 25`] = `
Number {
	Math.SQRT1_2
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 26`] = `
Number {
	Math.SQRT1_2
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 27`] = `
MathConstant {
	Math.SQRT1_2
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 28`] = `
MathConstant {
	Math.SQRT1_2
	
	bar: "Bar"
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 29`] = `
Number {
	Math.SQRT2
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 30`] = `
Number {
	Math.SQRT2
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 31`] = `
MathConstant {
	Math.SQRT2
}
`;

exports[`Boxed primitives Numbers identifies Math.* constants 32`] = `
MathConstant {
	Math.SQRT2
	
	bar: "Bar"
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 1`] = `
Number {
	Number.EPSILON
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 2`] = `
Number {
	Number.EPSILON
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 3`] = `
MagicNumber {
	Number.EPSILON
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 4`] = `
MagicNumber {
	Number.EPSILON
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 5`] = `
Number {
	Number.MIN_VALUE
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 6`] = `
Number {
	Number.MIN_VALUE
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 7`] = `
MagicNumber {
	Number.MIN_VALUE
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 8`] = `
MagicNumber {
	Number.MIN_VALUE
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 9`] = `
Number {
	Number.MAX_VALUE
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 10`] = `
Number {
	Number.MAX_VALUE
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 11`] = `
MagicNumber {
	Number.MAX_VALUE
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 12`] = `
MagicNumber {
	Number.MAX_VALUE
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 13`] = `
Number {
	Number.MIN_SAFE_INTEGER
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 14`] = `
Number {
	Number.MIN_SAFE_INTEGER
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 15`] = `
MagicNumber {
	Number.MIN_SAFE_INTEGER
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 16`] = `
MagicNumber {
	Number.MIN_SAFE_INTEGER
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 17`] = `
Number {
	Number.MAX_SAFE_INTEGER
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 18`] = `
Number {
	Number.MAX_SAFE_INTEGER
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 19`] = `
MagicNumber {
	Number.MAX_SAFE_INTEGER
}
`;

exports[`Boxed primitives Numbers identifies Number.* constants 20`] = `
MagicNumber {
	Number.MAX_SAFE_INTEGER
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies subclasses 1`] = `
Double {
	300
}
`;

exports[`Boxed primitives Numbers identifies subclasses 2`] = `
Double {
	300
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers identifies subclasses 3`] = `
Quadruple {
	600
}
`;

exports[`Boxed primitives Numbers identifies subclasses 4`] = `
Quadruple {
	600
	
	bar: "Bar"
}
`;

exports[`Boxed primitives Numbers prints internal values 1`] = `
Number {
	64
}
`;

exports[`Boxed primitives Numbers prints internal values 2`] = `
Number {
	-326.2
}
`;

exports[`Boxed primitives Numbers prints properties 1`] = `
Number {
	48
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Numbers prints properties 2`] = `
Number {
	48
	
	foo: "Foo"
	bar: "Bar"
}
`;

exports[`Boxed primitives Numbers resolves internal values with \`Number#valueOf()\` 1`] = `
Number {
	258
	
	valueOf: Function {
		│1│ function () {
		│2│         return 30;
		│3│       }
	}
}
`;

exports[`Boxed primitives Numbers resolves internal values with \`Number#valueOf()\` 2`] = `
Number {
	258
	
	valueOf: Function {
		│1│ function () {
		│2│         return 30;
		│3│       }
	}
	toString: Function {
		│1│ function () {
		│2│         return "982";
		│3│       }
	}
}
`;

exports[`Boxed primitives Numbers resolves internal values with \`Number#valueOf()\` 3`] = `
LyingNumber {
	32
}
`;

exports[`Boxed primitives Numbers resolves internal values with \`Number#valueOf()\` 4`] = `
LyingNumber {
	32
	
	foo: "Foo"
}
`;

exports[`Boxed primitives Strings escapes whitespace characters: "\\f" 1`] = `
String {
	"\\f"
	
	0: "\\f"
}
`;

exports[`Boxed primitives Strings escapes whitespace characters: "\\f" 2`] = `
String {
	"\\f"
	
	0: "\\f"
	foo: "Foo"
}
`;

exports[`Boxed primitives Strings escapes whitespace characters: "\\f" 3`] = `
Whitespace {
	"\\f"
	
	0: "\\f"
}
`;

exports[`Boxed primitives Strings escapes whitespace characters: "\\f" 4`] = `
Whitespace {
	"\\f"
	
	0: "\\f"
	bar: "Bar"
}
`;

exports[`Boxed primitives Strings escapes whitespace characters: "\\n" 1`] = `
String {
	"\\n"
	
	0: "\\n"
}
`;

exports[`Boxed primitives Strings escapes whitespace characters: "\\n" 2`] = `
String {
	"\\n"
	
	0: "\\n"
	foo: "Foo"
}
`;

exports[`Boxed primitives Strings escapes whitespace characters: "\\n" 3`] = `
Whitespace {
	"\\n"
	
	0: "\\n"
}
`;

exports[`Boxed primitives Strings escapes whitespace characters: "\\n" 4`] = `
Whitespace {
	"\\n"
	
	0: "\\n"
	bar: "Bar"
}
`;

exports[`Boxed primitives Strings escapes whitespace characters: "\\t" 1`] = `
String {
	"\\t"
	
	0: "\\t"
}
`;

exports[`Boxed primitives Strings escapes whitespace characters: "\\t" 2`] = `
String {
	"\\t"
	
	0: "\\t"
	foo: "Foo"
}
`;

exports[`Boxed primitives Strings escapes whitespace characters: "\\t" 3`] = `
Whitespace {
	"\\t"
	
	0: "\\t"
}
`;

exports[`Boxed primitives Strings escapes whitespace characters: "\\t" 4`] = `
Whitespace {
	"\\t"
	
	0: "\\t"
	bar: "Bar"
}
`;

exports[`Boxed primitives Strings identifies subclasses 1`] = `
ExtendedString {
	"XYZ"
	
	0: "X"
	1: "Y"
	2: "Z"
}
`;

exports[`Boxed primitives Strings identifies subclasses 2`] = `
ExtendedString {
	"XYZ"
	
	0: "X"
	1: "Y"
	2: "Z"
	foo: "Foo"
}
`;

exports[`Boxed primitives Strings prints internal values 1`] = `
String {
	"ABC"
	
	0: "A"
	1: "B"
	2: "C"
}
`;

exports[`Boxed primitives Strings prints properties 1`] = `
String {
	"XYZ"
	
	0: "X"
	1: "Y"
	2: "Z"
	foo: "Foo"
}
`;

exports[`Boxed primitives Strings resolves internal values with \`String#valueOf()\` 1`] = `
String {
	"ABC"
	
	0: "A"
	1: "B"
	2: "C"
	toString: Function {
		│1│ function () {
		│2│         return "XYZ";
		│3│       }
	}
	valueOf: Function {
		│1│ function () {
		│2│         return "Nah";
		│3│       }
	}
}
`;

exports[`Boxed primitives Strings resolves internal values with \`String#valueOf()\` 2`] = `
WeirdString {
	"ABC"
	
	0: "A"
	1: "B"
	2: "C"
}
`;
